src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php

```diff
@@ -274,12 +286,38 @@ private function getCategoriesInformation(Product $product): CategoriesInformati
 
         $categoriesInformation = [];
         foreach ($categoryNames as $categoryId => $localizedNames) {
-            $categoriesInformation[] = new CategoryInformation($categoryId, $localizedNames);
+            $localizedNames = $categoryNames[$categoryId];
+            $displayName = $this->buildDisplayName(
+                new CategoryId($categoryId),
+                $languageId,
+                $localizedNames[$this->contextLangId],
+                $duplicateNames
+            );
+            $categoriesInformation[] = new CategoryInformation(
+                $categoryId,
+                $displayName,
+                $localizedNames
+            );
         }
 
         return new CategoriesInformation($categoriesInformation, $defaultCategoryId);
     }
 
+    private function buildDisplayName(
+        CategoryId $categoryId,
+        LanguageId $languageId,
+        string $categoryName,
+        array $duplicateNames
+    ): string {
+        if (!in_array($categoryName, $duplicateNames)) {
+            return $categoryName;
+        }
+
+        $breadcrumbs = $this->categoryRepository->getBreadcrumbParts($categoryId, $languageId);
+
+        return implode(' > ', array_slice($breadcrumbs, -2, 2));
```

Comment:
Same here as in the tree handler this needs to be documented with much details because there's no way someone reaches this code and understands the purpose out of the blue